staging/src/k8s.io/apimachinery/pkg/runtime/helper.go

```diff
@@ -242,6 +243,14 @@ func (e WithVersionEncoder) Encode(obj Object, stream io.Writer) error {
 	return err
 }
 
+// CustomEncoder is an interface that allows intercepting encoding for
+// individual object to allow custom encoding.
+type CustomEncoder interface {
+	// InterceptEncode takes encoder that would have been used if the object
+	// didn't implement CustomEncoder interface.
+	InterceptEncode(encoder WithVersionEncoder, w io.Writer) error
```

Comment:
I'm not sure this interface is prescriptive enough. It exposes a lot of details about the encoder (presumably so the object can compute a cache key?), doesn't make it very clear what the object's responsibilities are, allows the called object to completely replace or modify the object passed to the encoder in cache miss cases, and also allows the called object to accidentally trigger an infinitely recursive call.

Rather than exposing details of the encoder, would it work to put the responsibility for computing a cache key on the encoder stack? I think a given encoder stack should be able to generate a cache key by accumulating the options of the delegates, like serialization (yaml, json, proto, etc), target version, etc.

Rather than implicitly relying on the interceptor to unwrap and pass back an encodeable object, should we make getting the encodeable object part of the interceptor interface?

```
type CacheableObject interface {
  ...
  GetEncodeableObject() runtime.Object
}
```

Reply:
for an object that implements this interface, what would it's implementation of DeepCopy() look like? would it preserve or discard the cached data?

how would we protect against this being accidentally used in code paths where objects are modified and serialized, and the modifications expected to be reflected in the serialized content?

how would we ensure that Encode() implementations that do type checks (for unstructured, etc) don't get bypassed by objects wrapped to implement caching/intercept interfaces?

Reply:
Couple notes to myself so I don't forget to discuss tomorrow.
1. I initially thought that this might be great because it would mean that if we forget to handle the CacheableObject interface, then it can only be performance hit. But I'm not really sure if we can make that work so that serializing that object will result always in exactly the same format as the raw object to be serialized.
2. We would need to extend the interface of Encoder (or Serializer) to be able to return its options that uniquely identify it (e.g. yaml/json/proto, pretty, GV (for versioning), and so on. Is that acceptable?
3. For now I think the GetEncodeableObject() would need to probably take those options as a parameter - SelfLink is the reason [we send a single CacheableObject to all watchers, and conversions to different versions result in different selflinks]. [It also need to make a DeepCopy, but that seems to match the model].
4. We should forbid deep-copying of that object.
5. Modifying object is tricky: we can implement our own accessor for metadata and control it (or at least know that something went wrong), but should we also care about the rest of object? Can those be modified by something that is not an encoder?
6. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go#L40 is what is tricky, because it's not even a serializer and it's manipulating objects

Reply:
Couple more thoughts:
1. I think (1) from above is still something I don't have any reasonable answer.
2. Regaring Jordan`s suggestion about the interface for `CacheableObject`, I'm now leaning toward the interface to be more like:
```
type CacheableObject interface {
  // The assumption is that for a given identifier we assume that if called multiple times, 
  // encode function will always be the same.
  CacheEncode(id SerializerIdentifier, encode func(runtime.Object, io.Writer)) ([]byte, error)
  // Probably also
  GetObject(id SerializerIdentifier) runtime.Object
```

The first makes it easy to use, because we simply pass some identifier, and encoder.Encode() function as parameters.
The second (in some form) will probably be needed to handle places like https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go#L40 (it seems that currently this is the only place, but who knows how it will look like in the future).

Reply:
Objects (like unstructured) could implement CacheableObject directly, and wouldn't have to unwrap themselves when invoking encode()

Reply:
```
  // Probably also
  GetObject(id SerializerIdentifier) runtime.Object
```

finding all the places that do typechecks and making sure they call this will be tricky. also, we shouldn't vary the encodeable object based on the serializer id, so probably no param is needed

Reply:
```
CacheEncode(id SerializerIdentifier, encode func(runtime.Object, io.Writer)) ([]byte, error)
```

rather than getting back []byte, can we propagate the writer down so the object can stream to it